### DISCUSSION TITLE
Network time synchronization

### DISCUSSION THREAD

**Discussion by anzas** [openthread/openthread#8752] [Q&A]
I have tested the network time synchronization feature and I got it somewhat working with quite good accuracy, but there are few limitations I would like to ask about. 
It seems that the leader time sync is actually not the same, when asked with the otNetworkTimeGet(). Should the leader time be treated otherwise? 
To acquire precise time sync I need to set the time sync period to 5 seconds. This makes the time base change quite drastically on every sync. Is this how it should behave? All child devices have the same time, it only changes thousands of seconds on every sync.
Lastly, it seems that the sync is not transferred from router to child. I only get it working in a network where all the devices are connected to leader. Should the time sync work throughout the network?

**Comment by jwhui** [comment]
Note that the Network Time sync feature is experimental and not a part of the Thread Specification. Enabling the Network Time feature may break compatibility with Thread conformant devices.

> It seems that the leader time sync is actually not the same, when asked with the otNetworkTimeGet(). Should the leader time be treated otherwise?

The `otNetworkTimeGet() value should be (roughly) the same across the network when time is synchronized.

> To acquire precise time sync I need to set the time sync period to 5 seconds. This makes the time base change quite drastically on every sync. Is this how it should behave? All child devices have the same time, it only changes thousands of seconds on every sync.

If working properly, the time sync value should only change within the crystal accuracy parameters.

> Lastly, it seems that the sync is not transferred from router to child. I only get it working in a network where all the devices are connected to leader. Should the time sync work throughout the network?

Network Time should propagate across routers as well. Note that the Network Time implementation requires all devices on the network to enable this feature in order to properly distribute and synchronize time.

**Comment by anzas** [comment]
Thanks for the answer. I have been able to add timesync traces and have discovered that the nodes receive a lot of sync messages with irrelevant sequence numbers. From the beginning the leader sends sequence 0, which is received by the router. After a while the router receives sequence 224 and then 117, but no traces of who sent those. There are only three devices in the network and I have been tracing all of them. I guess the inconsistent time on router and child device is caused by these sync messages. Do you have any suggestions on how to debug the problem.

**Reply by jwhui** [reply]
The leader should be authoritative in incrementing the sequence number. All other devices should simply adopt the larger timestamp value as they are received. I would try to track down whether nodes are properly receiving the sequence number values and repeating them in their own transmissions.
